### PR TITLE
small fix for activation layer to work with stft layer using absolute activation

### PR DESCRIPTION
**Pull request by dierkes-j**
When using the Activation Layer with absolute activation for the output of a stft layer there is an assertion failure that the output should have type "float32" but actually has type "complex64". I propose a fix of the problem and a test for its correctness.

### CODE REVIEW

File: returnn/tf/layers/basic.py [hunk 1]
```diff
@@ -550,6 +550,9 @@ def __init__(self, activation, **kwargs):
     if self.output_before_activation:
       self.output.placeholder = self.output_before_activation.y
 
+    if activation == "abs" and self.output.dtype == "complex64":
+      self.output.dtype = "float32"
```
Comment (albertz) [inline returnn/tf/layers/basic.py:554]: This should not be part of `__init__` but rather in `get_out_data_from_opts`.

Comment (albertz) [reply returnn/tf/layers/basic.py:554]: Actually, it is already. So I don't understand, why is this needed?

File: tests/test_TFNetworkLayer.py [hunk 2]
```diff
@@ -388,6 +388,40 @@ def test_activation_layer_net_construct():
...
+      "num_outputs": 3,
+      "num_inputs": num_inputs,
+      "network": {
+        "stft": { "class": "multichannel_stft_layer", "from": "data",
+                  "frame_shift": frame_shift, "frame_size": frame_size,
+                  "window": "hanning", "fft_size": fft_size,
+                  "use_rfft": True, "pad_last_frame": False,
+                  "nr_of_channels": 1 },
+        "output": { "class": "activation", "activation": "abs", "from": ["stft"] }
+      }})
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py:408]: Please use correct PEP8 formatting.

File: returnn/tf/layers/basic.py (hunk 1 above)
Comment (albertz) [reply returnn/tf/layers/basic.py:554]: I.e. that means that we should change the base to `LayerBase`.

Comment (dierkes-j) [reply returnn/tf/layers/basic.py]: I agree that we should do things the right and clean way, I just did not know better. Are you sure we should use the `LayerBase`? I updated my fix using the `_ConcatInputLayer` as the base, then concatination of input is still possible and we do not have to change the existing logic in the `ActivationLayer's` `__init__` method.

Comment (albertz) [reply returnn/tf/layers/basic.py]: Ah right, `_ConcatInputLayer` makes more sense then.

File: returnn/tf/layers/basic.py [hunk 3]
```diff
@@ -551,19 +551,18 @@ def __init__(self, activation, **kwargs):
       self.output.placeholder = self.output_before_activation.y
 
   @classmethod
-  def get_out_data_from_opts(cls, activation, **kwargs):
+  def get_out_data_from_opts(cls, sources, name, activation, **kwargs):
```
Comment (albertz) [inline returnn/tf/layers/basic.py:554]: The new arguments should also be documented.

File: tests/test_TFNetworkLayer.py [hunk 4]
```diff
@@ -388,6 +388,40 @@ def test_activation_layer_net_construct():
     assert_equal(v.shape, (n_batch, seq_len, num_inputs))
     assert_equal(v.tolist(), [[[0, 0], [0, 0], [2, 2]]])
 
+def test_activation_layer_abs_for_stft():
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py:391]: Add two empty lines before a function.
(Please check our generic code style rules.)

Comment (JackTemaki) [reply tests/test_TFNetworkLayer.py:391]: https://github.com/rwth-i6/returnn/wiki/PyCharm-Configuration

Might help you with this

Comment (dierkes-j) [reply tests/test_TFNetworkLayer.py:392]: I corrected the emtpy lines.  Thank you Nick, I adjusted my project configuration.

File: returnn/tf/layers/basic.py (hunk 3 above)
Comment (dierkes-j) [reply returnn/tf/layers/basic.py:554]: You are right, I added the documentations.

File: returnn/tf/layers/basic.py [hunk 5]
```diff
@@ -551,19 +551,20 @@ def __init__(self, activation, **kwargs):
...
-    out = super(ActivationLayer, cls).get_out_data_from_opts(**kwargs)
-    # Modify if needed based on activation function
+    # get dtype based on the inputs
+    out = get_concat_sources_data_template(sources, name="%s_output" % name)
+    # modify if needed based on activation function
     if activation == "abs" and out.dtype == "complex64":
       out.dtype = "float32"
     return out
 
-
```
Comment (albertz) [inline returnn/tf/layers/basic.py:566]: Why was this empty line removed? It is needed here.

Comment (dierkes-j) [reply returnn/tf/layers/basic.py]: I dont know, I confused myself.

**Review by albertz — APPROVED** [review]

**Comment by albertz** [comment]
There is some test (`test_engine_rec_subnet_count`) failing.
```
  File "/home/runner/work/returnn/returnn/returnn/tf/network.py", line 965, in TFNetwork._create_layer
    line: layer = layer_class(**layer_desc)
    locals:
      layer = <not found>
      layer_class = <local> <class 'returnn.tf.layers.basic.ActivationLayer'>
      layer_desc = <local> {'activation': 'identity + 1', 'initial_output': 0, 'out_type': {'dim': 1, 'dtype': 'int32'}, '_network': <TFNetwork 'root/output(rec-subnet)' parent_layer=<RecLayer 'output' out_type=Data(batch_shape_meta=[T|F|'time:var:extern_data:data',B])> train=<tf.Tensor 'globals/train_flag:0' shape=() dtyp..., len = 10
  File "/home/runner/work/returnn/returnn/returnn/tf/layers/basic.py", line 536, in ActivationLayer.__init__
    line: super(ActivationLayer, self).__init__(**kwargs)
    locals:
      super = <builtin> <class 'super'>
      ActivationLayer = <global> <class 'returnn.tf.layers.basic.ActivationLayer'>
      self = <local> <ActivationLayer output/'output' out_type=Data(batch_shape_meta=[B])>
      __init__ = <not found>
      kwargs = <local> {'initial_output': 0, 'out_type': {'dim': 1, 'dtype': 'int32'}, '_network': <TFNetwork 'root/output(rec-subnet)' parent_layer=<RecLayer 'output' out_type=Data(batch_shape_meta=[T|F|'time:var:extern_data:data',B])> train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>, '_name': 'output', '..., len = 9
  File "/home/runner/work/returnn/returnn/returnn/tf/layers/basic.py", line 228, in _ConcatInputLayer.__init__
    line: super(_ConcatInputLayer, self).__init__(**kwargs)
    locals:
      super = <builtin> <class 'super'>
      _ConcatInputLayer = <global> <class 'returnn.tf.layers.basic._ConcatInputLayer'>
      self = <local> <ActivationLayer output/'output' out_type=Data(batch_shape_meta=[B])>
      __init__ = <not found>
      kwargs = <local> {'initial_output': 0, 'out_type': {'dim': 1, 'dtype': 'int32'}, '_network': <TFNetwork 'root/output(rec-subnet)' parent_layer=<RecLayer 'output' out_type=Data(batch_shape_meta=[T|F|'time:var:extern_data:data',B])> train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>, '_name': 'output', '..., len = 9
  File "/home/runner/work/returnn/returnn/returnn/tf/layers/base.py", line 149, in LayerBase.__init__
    line: assert self.output.dim == out_type["dim"]
    locals:
      self = <local> <ActivationLayer output/'output' out_type=Data(batch_shape_meta=[B])>
      self.output = <local> Data(name='output_output', batch_shape_meta=[B])
      self.output.dim = <local> None
      out_type = <local> {'dim': 1, 'dtype': 'int32'}
```
```
Exception creating layer root/output(rec-subnet)/'output' of class ActivationLayer with opts:
{'_name': 'output',
 '_network': <TFNetwork 'root/output(rec-subnet)' parent_layer=<RecLayer 'output' out_type=Data(batch_shape_meta=[T|F|'time:var:extern_data:data',B])> train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>,
 'activation': 'identity + 1',
 'initial_output': 0,
 'name': 'output',
 'network': <TFNetwork 'root/output(rec-subnet)' parent_layer=<RecLayer 'output' out_type=Data(batch_shape_meta=[T|F|'time:var:extern_data:data',B])> train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>,
 'out_type': {'dim': 1, 'dtype': 'int32'},
 'output': Data(name='output_output', batch_shape_meta=[B]),
 'rec_previous_layer': <_TemplateLayer(ActivationLayer)(:prev:activation) output/'prev:output' out_type=Data(batch_shape_meta=[B?]) (construction stack None)>,
 'sources': [<_TemplateLayer(ActivationLayer)(:prev:activation) output/'prev:output' out_type=Data(batch_shape_meta=[B?]) (construction stack None)>]}
```

File: returnn/tf/layers/basic.py [hunk 6]
```diff
@@ -551,14 +551,16 @@ def __init__(self, activation, **kwargs):
...
+    :param str name:
     :param str activation:
     :rtype: Data
     """
-    # Get dtype based on inputs
-    out = super(ActivationLayer, cls).get_out_data_from_opts(**kwargs)
-    # Modify if needed based on activation function
+    # get dtype based on the inputs
+    out = get_concat_sources_data_template(sources, name="%s_output" % name)
+    # modify if needed based on activation function
```
Comment (albertz) [inline returnn/tf/layers/basic.py:563]:
```suggestion
    # Just the same as the input.
    # Use CopyLayer.get_out_data_from_opts for potential extra logic for out_type.
    out = CopyLayer.get_out_data_from_opts(**kwargs)
    # Modify dtype if needed based on activation function
```

File: returnn/tf/layers/basic.py [hunk 7]
```diff
@@ -551,14 +551,16 @@ def __init__(self, activation, **kwargs):
       self.output.placeholder = self.output_before_activation.y
 
   @classmethod
-  def get_out_data_from_opts(cls, activation, **kwargs):
+  def get_out_data_from_opts(cls, sources, name, activation, **kwargs):
     """
+    :param list[LayerBase] sources:
+    :param str name:
```
Comment (albertz) [inline returnn/tf/layers/basic.py:557]: Sorry for the back and forth, but with my new suggestion, this is not needed anymore.

```suggestion
  def get_out_data_from_opts(cls, activation, **kwargs):
    """
```

**Review by albertz — APPROVED** [review]

**Comment by albertz** [comment]
I committed another change to fix the failing test. And merged now.

**Pull request merged**